megatron/data/dataset_utils.py

```diff
@@ -267,15 +338,25 @@ def create_masked_lm_predictions(tokens,
                 if index in covered_indexes:
                     continue
 
-        if not geometric_dist:
+        if sampling_style is SamplingStyle.POISSON:
             n = np_rng.choice(ngrams[:len(cand_index_set)],
                               p=pvals[:len(cand_index_set)] /
                               pvals[:len(cand_index_set)].sum(keepdims=True))
-        else:
+        elif sampling_style is SamplingStyle.GEOMETRIC:
             # Sampling "n" from the geometric distribution and clipping it to
             # the max_ngrams. Using p=0.2 default from the SpanBERT paper
             # https://arxiv.org/pdf/1907.10529.pdf (Sec 3.1)
             n = min(np_rng.geometric(0.2), max_ngrams)
+        elif sampling_style is SamplingStyle.UNIFORM:
+            n = np_rng.choice(ngrams[:len(cand_index_set)])
+        elif sampling_style is SamplingStyle.NORMAL:
+            n = round(np.clip(
+                np_rng.normal(loc=normal_mean),
```

Comment:
This normal distribution has a standard deviation of 1 by default.
I didn't find any information about that in the UL2 paper, but would it make sense to scale that with the mean instead? For example `scale=np.sqrt(normal_mean)`

Reply:
I don't think it's desired for the spans to vary in length that heavily. You did make me look up what's happening in the `seqio` implementation, though. At first glance, it seems like they're actually using a uniform distribution after all. This was really only a cursory glance, take this with a grain of salt.
See https://github.com/google-research/text-to-text-transfer-transformer/blob/f0cf9e8c51bd48699265763d01c2f8b24ae1098b/t5/data/preprocessors.py#L2958-L2960
and https://github.com/google/seqio/blob/cd5bb07f22d3b36c7c10e906ebed97ab3efd780f/seqio/utils.py#L613-L620.

Reply:
Ahh that seqio code looks indeed very different from a normal distribution.
In Nemo they are using `scale=np.sqrt(normal_mean)` https://github.com/NVIDIA/NeMo/blob/989b07a0e87979a73572d4c37d2f19538eac9d16/nemo/collections/nlp/data/language_modeling/megatron/dataset_utils.py#L465

Reply:
That makes sense, thanks!

Reply:
It completely eluded me that UL2R uses very different denoiser settings and also describes their implementation differently from UL2 (e.g. uniform sampling of span lengths for R- and X-denoisers). I'll adjust the implementation to handle these.